File: 0328-odd-even-linked-list/0328-odd-even-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def oddEvenList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        evenNode = ListNode(0)
        oddNode = ListNode(1)
        evenHead = evenNode
        odddHead = oddNode
        isOdd = True
        while head:
            if isOdd:
                evenNode.next = head
                evenNode = evenNode.next
            else:
                oddNode.next = head
                oddNode = oddNode.next
            isOdd = not isOdd
            head = head.next
        oddNode.next = None
        evenNode.next = odddHead.next
        return evenHead.next

File: 0328-odd-even-linked-list/0328-odd-even-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode oddEvenList(ListNode head) {
        if (head == null || head.next == null) {
            return head;
        }
        ListNode evenList = new ListNode();
        ListNode oddList = new ListNode();
        ListNode evenHead = evenList;
        ListNode oddHead = oddList;
        ListNode dummy = head;
        int index = 0;

        while (dummy != null) {
            ListNode next = dummy.next;
            dummy.next = null;
            if (index % 2 == 0) {
                evenHead.next = dummy;
                evenHead = evenHead.next;
            } else {
                oddHead.next = dummy;
                oddHead = oddHead.next;
            }
            index++;
            dummy = next;
        }
        evenHead.next = oddList.next;
        return evenList.next;
        
        
    }
}